controllers/core/openstackcontrolplane_controller.go

```diff
@@ -245,6 +260,59 @@ func (r *OpenStackControlPlaneReconciler) reconcileNormal(ctx context.Context, i
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackControlPlaneReconciler) reconcileDelete(ctx context.Context, instance *corev1beta1.OpenStackControlPlane, helper *helper.Helper) (ctrl.Result, error) {
+	var err error
+	overallRes := ctrl.Result{}
+
+	//
+	// Ensure a controlled teardown of the control plane...
+	//
+
+	// Ordered list of resource types to delete
+	// Each outer index of resource groupings must be fully deleted before the next index is considered
+	steps := [][]client.ObjectList{
+		{
+			&cinderv1.CinderList{},
+			&glancev1.GlanceList{},
+			&neutronv1.NeutronAPIList{},
+			&novav1.NovaList{},
+			&ovnv1.OVNDBClusterList{},
+			&ovnv1.OVNNorthdList{},
+			&ovsv1.OVSList{},
+			&placementv1.PlacementAPIList{},
+		},
+		{
+			&rabbitmqv1.RabbitmqClusterList{},
+			&keystonev1.KeystoneAPIList{},
+		},
+		{
+			&mariadbv1.MariaDBList{},
+		},
+	}
+
+	for _, step := range steps {
+		if overallRes, err = openstack.CheckDeleteSubresources(ctx, helper, step); err != nil {
```

Comment:
Since we are checking all subresources here does that imply we could be "lazy" and only rely on setting the ownership of subresources in the service operators to do actual cleanup. Versus adding implicit finalizers in each operator and having logic in each of the service operators similar to this.

Reply:
we spoke on slack. what I'm suggesting here would require perhaps extra fancy logic to guard against some of the cross/operator teardown steps. Like if mariadb disappears first it would cause some resources to hang and they would need to realize mariadb was gone and skip the finalizer. I think what we have here is fine.